File: golang/server/handlers/revenue.go
package handlers

import (
	"encoding/json"
	"net/http"

	"github.com/aituglo/rubyx/golang/db"
	"github.com/aituglo/rubyx/golang/env"
	"github.com/aituglo/rubyx/golang/errors"
	"github.com/aituglo/rubyx/golang/server/write"
)

func CreateRevenue(env env.Env, user *db.User, w http.ResponseWriter, r *http.Request) http.HandlerFunc {
	if user.Status != db.UserStatusActive {
		return write.Error(errors.RouteUnauthorized)
	}

	decoder := json.NewDecoder(r.Body)
	p := &db.Revenue{}
	err := decoder.Decode(p)
	if err != nil || p == nil {
		return write.Error(errors.NoJSONBody)
	}

	return write.JSONorErr(env.DB().CreateRevenue(r.Context(), db.CreateRevenueParams{
		ProgramID:       p.ProgramID,
		VulnerabilityID: p.VulnerabilityID,
		Money:           p.Money,
	}))
}

func GetRevenue(env env.Env, user *db.User, w http.ResponseWriter, r *http.Request) http.HandlerFunc {
	if user.Status != db.UserStatusActive {
		return write.Error(errors.RouteUnauthorized)
	}

	id, err := getID(r)
	if err != nil {
		return write.Error(errors.RouteNotFound)
	}

	revenue, err := env.DB().FindRevenueByIDs(r.Context(), id)
	if err != nil {
		if isNotFound(err) {
			return write.Error(errors.PostNotFound)
		}
		return write.Error(err)
	}

	return write.JSON(revenue)
}

func GetRevenues(env env.Env, user *db.User, w http.ResponseWriter, r *http.Request) http.HandlerFunc {
	if user.Status != db.UserStatusActive {
		return write.Error(errors.RouteUnauthorized)
	}

	return write.JSONorErr(env.DB().FindRevenues(r.Context()))
}

func UpdateRevenue(env env.Env, user *db.User, w http.ResponseWriter, r *http.Request) http.HandlerFunc {
	if user.Status != db.UserStatusActive {
		return write.Error(errors.RouteUnauthorized)
	}

	decoder := json.NewDecoder(r.Body)
	p := &db.Revenue{}
	err := decoder.Decode(p)
	if err != nil || p == nil {
		return write.Error(errors.NoJSONBody)
	}

	return write.JSONorErr(env.DB().UpdateRevenue(r.Context(), db.UpdateRevenueParams{
		ID:              p.ID,
		ProgramID:       p.ProgramID,
		VulnerabilityID: p.VulnerabilityID,
		Money:           p.Money,
	}))
}

func DeleteRevenue(env env.Env, user *db.User, w http.ResponseWriter, r *http.Request) http.HandlerFunc {
	if user.Status != db.UserStatusActive {
		return write.Error(errors.RouteUnauthorized)
	}

	id, err := getID(r)
	if err != nil {
		return write.Error(errors.RouteNotFound)
	}

	return write.SuccessOrErr(env.DB().DeleteRevenueByIDs(r.Context(), id))
}

File: dashboard/src/actions/revenue.js
import API from "./api";
import {
  GET_REVENUE,
  GET_REVENUE_ERROR,
  CREATE_REVENUE,
  CREATE_REVENUE_ERROR,
  DELETE_REVENUE,
  DELETE_REVENUE_ERROR,
  UPDATE_REVENUE,
  UPDATE_REVENUE_ERROR
} from "../constants/revenue";

export const getRevenues = () => async (dispatch) => {
  try {
    API.getRevenues()
      .then((data) => {
        dispatch({
          type: GET_REVENUE,
          payload: data,
        });
      })
      .catch((err) => {
        dispatch({
          type: GET_REVENUE_ERROR,
          payload: err,
        });
      });
  } catch (err) {
    dispatch({
      type: GET_REVENUE_ERROR,
      payload: err,
    });
  }
};

export const createRevenue = (program_id, vulnerability_id, money) => async (dispatch) => {
  try {
    var revenue = {
      program_id: program_id,
      vulnerability_id: vulnerability_id,
      money: money
    };

    API.createRevenue(revenue)
      .then((res) => {
        API.getRevenues().then((data) => {
          dispatch({
            type: CREATE_REVENUE,
            payload: data,
          });
        });
      })
      .catch((err) => {
        dispatch({
          type: CREATE_REVENUE_ERROR,
          payload: err,
        });
      });
  } catch (err) {
    dispatch({
      type: CREATE_REVENUE_ERROR,
      payload: err,
    });
  }
};

export const updateRevenue = (id, program_id, vulnerability_id, money) => async (dispatch) => {
  try {
    var revenue = {
      id: id,
      program_id: program_id,
      vulnerability_id: vulnerability_id,
      money: money
    };

    console.log(revenue)

    API.updateRevenue(revenue)
      .then((res) => {
        API.getRevenues().then((data) => {
          dispatch({
            type: UPDATE_REVENUE,
            payload: data,
          });
        });
      })
      .catch((err) => {
        dispatch({
          type: UPDATE_REVENUE_ERROR,
          payload: err,
        });
      });
  } catch (err) {
    dispatch({
      type: UPDATE_REVENUE_ERROR,
      payload: err,
    });
  }
};

export const deleteRevenue = (id) => async (dispatch) => {
  try {
    API.deleteRevenue(id)
      .then((res) => {
        API.getRevenues().then((data) => {
          dispatch({
            type: DELETE_REVENUE,
            payload: data,
          });
        });
      })
      .catch((err) => {
        dispatch({
          type: DELETE_REVENUE_ERROR,
          payload: err,
        });
      });
  } catch (err) {
    dispatch({
      type: DELETE_REVENUE_ERROR,
      payload: err,
    });
  }
};
